### PR TITLE
[KYUUBI #6247] Make KSHC binary compatible with multiple Spark versions

### LINKED ISSUE #6247: [TASK][MEDIUM] Make KSHC binary compatible with multiple Spark versions
### What's the level of this task?

MEDIUM

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before creating

- [X] I have searched in the [task list](https://github.com/orgs/apache/projects/343) and found no similar tasks.

### Mentor

- [X] I have sufficient expertise on this task, and I volunteer to be a mentor of this task to guide contributors through the task.

### Skill requirements

- Knowledge of GitHub Actions, Java Reflections, Hive, Spark

### Background and Goals

Currently, the KSHC(Kyuubi Spark Hive Connector) supports Spark 3.3, 3.4, and 3.5, but unlike the Spark engine, the KSHC jar built with Spark 3.5 can not work on Spark 3.4 runtime.

### Implementation steps

- Tune the KSHC code to fix binary-compatible issues
- Tune the test code and set up the GitHub Actions workflow to do cross-version verification

### Additional context

Introduction of [2024H1 Kyuubi Code Contribution Program](https://github.com/apache/kyuubi/issues/6232)

### LINKED ISSUE #6431: [FEATURE] [spark-hive-connector] support scala 2.13 for spark 3.3 version
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/kyuubi/issues?q=is%3Aissue) and found no similar issues.

### Describe the feature

support scala 2.13 for spark 3.3 version

### Motivation

It's hard for downstream code using Kyuubi spark hive connector like `Gravitino` to support multi scala versions in https://github.com/datastrato/gravitino/pull/3608

### Describe the solution

compile with scala 2.13 and publish to Maven

### Additional context

_No response_

### Are you willing to submit PR?

- [X] Yes. I would be willing to submit a PR with guidance from the Kyuubi community to improve.
- [ ] No. I cannot submit a PR at this time.

### PR DESCRIPTION
**Pull request by zhouyifan279**
# :mag: Description
## Issue References 🔗

This pull request closes #6247

This also closes #6431

## Describe Your Solution 🔧
Add a job `spark-connector-cross-version-test` in GitHub Actions to:
1. Build KSHC package with maven opt `-Pspark-3.5`
2. Run KSHC tests with maven opt `-Pspark-3.3` and `-Pspark-3.4` and KSHC package built in step 1
3. Fix the binary-compatible issue via reflection.

## Types of changes :bookmark:

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Test Plan 🧪

Pass GHA.

---

# Checklist 📝

- [x] This patch was not authored or co-authored using [Generative Tooling](https://www.apache.org/legal/generative-tooling.html)

**Be nice. Be informative.**

### CODE REVIEW

File: .github/workflows/master.yml [hunk 1]
```diff
@@ -170,6 +170,54 @@ jobs:
             **/kyuubi-jdbc-engine.log*
             **/kyuubi-hive-sql-engine.log*
 
+  spark-connector-cross-version-test:
+    name: Kyuubi Spark Connector Cross Version Test
+    runs-on: ubuntu-22.04
+    strategy:
+      fail-fast: false
+      max-parallel: 6
```
Comment (pan3793) [inline .github/workflows/master.yml:178]: please limit it to 1, we got asked to 

> All workflows SHOULD have a job concurrency level less than or equal to 15.

https://infra.apache.org/github-actions-policy.html

File: .github/workflows/master.yml [hunk 2]
```diff
@@ -170,6 +170,54 @@ jobs:
...
+      matrix:
+        java:
```
Comment (pan3793) [inline .github/workflows/master.yml:180]: might be good to use Java 17 only, Scala version should be considered too.

File: .github/workflows/master.yml (hunk 1 above)
Comment (zhouyifan279) [reply .github/workflows/master.yml]: Good to know. Updated.

File: .github/workflows/master.yml (hunk 2 above)
Comment (zhouyifan279) [reply .github/workflows/master.yml:180]: Updated.

**Comment by zhouyifan279** [comment]
@pan3793 added code to fix compatible issue. Please take another review.

**Comment by zhouyifan279** [comment]
cc @SteNicholas

File: .github/workflows/master.yml [hunk 3]
```diff
@@ -170,6 +170,61 @@ jobs:
...
+        with:
+          distribution: temurin
+          java-version: ${{ matrix.java }}
+          cache: 'maven'
+          check-latest: false
+      - name: Setup Maven
+        uses: ./.github/actions/setup-maven
+      - name: Build Kyuubi Spark Hive Connector with Spark-${{ matrix.spark-compile }}
+        run: |
+          ./build/mvn clean install ${MVN_OPT} -pl extensions/spark/kyuubi-spark-connector-hive -am \
```
Comment (pan3793) [inline .github/workflows/master.yml:210]: The TPC-DS/H connectors should be compatible too, can we add them to the x-version test too?

**Comment by pan3793** [comment]
Well done, cc @FANNG1 too

**Comment by pan3793** [comment]
Merged to master/1.9, and we can do TPC-DS/H connector cross verification later

File: .github/workflows/master.yml (hunk 3 above)
Comment (zhouyifan279) [reply .github/workflows/master.yml:210]: Sure. I'll do it in another PR.

**Comment by FANNG1** [comment]
> Well done, cc @FANNG1 too

cool, Is there any plan to publish the package supports Spark3.3 and Scala 2.13 to maven?

**Pull request closed without merging**